server/src/main/java/com/cloud/server/ManagementServerImpl.java

```diff
@@ -2392,7 +2392,12 @@ public String getConsoleAccessUrlRoot(final long vmId) {
             s_logger.trace("Trying to retrieve VNC port from agent about VM " + vm.getHostName());
         }
 
-        final GetVncPortAnswer answer = (GetVncPortAnswer)_agentMgr.easySend(vm.getHostId(), new GetVncPortCommand(vm.getId(), vm.getInstanceName()));
+        GetVncPortAnswer answer = null;
+        if (vm.getState() == State.Migrating && vm.getLastHostId() != null) {
+            answer = (GetVncPortAnswer)_agentMgr.easySend(vm.getLastHostId(), new GetVncPortCommand(vm.getId(), vm.getInstanceName()));
+        } else {
```

Comment:
So we are making the decision we are opening a console to the old VM. looks good but may have some problems. I suspect you tested @ustcweizhou .
problems I see are:
- this is a second migrate and the old host is no longer valid.
- the migration is already successful but not administered yet.

Reply:
@DaanHoogland indeed there are some problems with vm console, for example
(1) vm is Starting state, but is not running on any server  (eg applying dhcp/dns entries in virtual routers)
(2) vm is Stopping state, but is stopped on server
(3) vm is Migrating state, when we open the console of vm (on source host), and migration is done, the console is invalid.
However, I think all these problems are not real issues. When vm state is updated, refresh the web page and a new console will work.

Reply:
I agree, @weizhouapache these are not big issues. I just wanted to have mentioned them and am not 👎ing this!